---
title: "From raw reads to ASVs"
output:
 html_document:
        toc: true
        toc_float: true
        code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Here, we describe the the bioinformatics processing of the elephant seal
gut microbiome samples. We started from Illumina-sequenced paired-end fastq files 
and ended with an amplicon sequence variant (ASV) table. We will largely follow
the [DADA2 pipeline](https://benjjneb.github.io/dada2/tutorial.html).


```{r, message = FALSE, warning = FALSE }
library(dplyr)
library(dada2)
library(magrittr)
```

Create a folder for the output.
```{r}
# output folder
output_folder <- "primer_clipped_reads_11_220230_pool"
dir.create(paste0(paste0("output/", output_folder)))
```

Next, we define two of the parameters for the pipeline. ees defines how 
many errors are allowed and totrim specifies where reads should be trimmed.
```{r}
# filtering
ees <- c(1,1)
# trimming
totrim <- c(220,230) 
```





